File: node/be/api/steps/security.steps.js
'use strict';

const { When, Then } = require('@cucumber/cucumber');
const { MiniBooks, ApiUnreachable } = require('../venues/books');

/**
 * Steps for features/be-mini-books-security.feature. These probe the API's
 * authorization boundaries -- no token, a forged token, the wrong role. Setup
 * Givens (an accountant, a cash sale) and "the response status is {int}" are
 * shared from the other @books steps; only the adversarial requests and the
 * secret-hygiene assertions are new.
 */

const books = new MiniBooks();
const FORGED = 'usr_forged000000000000000000';
const twoLine = (dr, cr, amt) => [{ account_id: dr, side: 'debit', amount_cents: amt }, { account_id: cr, side: 'credit', amount_cents: amt }];

async function send(world, method, path, opts = {}) {
  if (world.sourceError) return;
  try { world.api = await books.request(method, path, opts); } catch (err) { if (err instanceof ApiUnreachable) { world.sourceError = err.message; return; } throw err; }
}
function check(world, description, fn) {
  if (world.sourceError) { world.unobservable(description, 'the source could not be reached -- ' + world.sourceError); return; }
  const r = fn(); world.check(description, r.passed, r.detail);
}
const j = (v) => JSON.stringify(v);
function leaksToken(obj, adminToken) { const text = j(obj || {}); return /"usr_[A-Za-z0-9_-]{8,}"/.test(text) || text.includes('"' + adminToken + '"'); }

// ---------------------------------------------------------------- adversarial requests

When('an entry is posted with no token', async function () { await send(this, 'POST', '/journal-entries', { body: { entry_date: books.bizDay, lines: twoLine(1, 9, 1000) } }); });
When('an entry is posted with a forged token', async function () { await send(this, 'POST', '/journal-entries', { token: FORGED, body: { entry_date: books.bizDay, lines: twoLine(1, 9, 1000) } }); });
When('entry {int} is read with no token', async function (id) { await send(this, 'GET', '/journal-entries/' + id); });
When('the admin entry list is read with the accountant\'s token', async function () { await send(this, 'GET', '/admin/entries', { token: books.seedBook }); });
When('the admin overview is read with the accountant\'s token', async function () { await send(this, 'GET', '/admin/overview', { token: books.seedBook }); });
When('the admin overview is read with a token that extends the admin token', async function () { await send(this, 'GET', '/admin/overview', { token: books.adminToken + 'x' }); });
When('the trial balance is read with no token', async function () { await send(this, 'GET', '/reports/trial-balance'); });
When('the chart of accounts is read with no token', async function () { await send(this, 'GET', '/accounts'); });
When('an accountant registers', async function () { await send(this, 'POST', '/users', { body: { name: 'Sec Accountant' } }); });

// ---------------------------------------------------------------- generic assertions

Then('the response is an error with code {string}', function (code) {
  check(this, 'error code ' + code, () => ({ passed: !!this.api && this.api.body && this.api.body.code === code, detail: this.api && this.api.body ? j(this.api.body) : 'no body' }));
});
Then('the response carries a token', function () {
  check(this, 'response carries a token', () => { const t = (this.api && this.api.body || {}).token; return { passed: typeof t === 'string' && t.length > 0, detail: 'token ' + (t ? 'present' : 'absent') }; });
});

// ---------------------------------------------------------------- secret hygiene

Then('the entry response carries no bearer token', function () { check(this, 'entry body has no token', () => ({ passed: !leaksToken(this.entry, books.adminToken), detail: leaksToken(this.entry, books.adminToken) ? 'TOKEN LEAKED' : 'clean' })); });
Then('the accounts response carries no bearer token', function () { check(this, 'accounts body has no token', () => ({ passed: !leaksToken(this.api && this.api.body, books.adminToken), detail: leaksToken(this.api && this.api.body, books.adminToken) ? 'TOKEN LEAKED' : 'clean' })); });
Then('the report response carries no bearer token', function () { check(this, 'report body has no token', () => ({ passed: !leaksToken(this.api && this.api.body, books.adminToken), detail: leaksToken(this.api && this.api.body, books.adminToken) ? 'TOKEN LEAKED' : 'clean' })); });

File: python/be/api/steps/security_steps.py
"""Steps for be-mini-books-security.feature. Mirror of node/be/api/steps/security.steps.js
-- a plugin module. These probe the API's authorization boundaries -- no token, a
forged token, the wrong role. Setup Givens (an accountant, a cash sale) and "the
response status is {int}" are shared from the other @books steps; only the adversarial
requests and the secret-hygiene assertions are new. qa.api is always the last response.
"""

from __future__ import annotations

import json
import re

from pytest_bdd import parsers, then, when

from be.api.venues.books import ApiUnreachable, MiniBooks

books = MiniBooks()
FORGED = "usr_forged000000000000000000"


def _two_line(dr, cr, amt):
    return [{"account_id": dr, "side": "debit", "amount_cents": amt}, {"account_id": cr, "side": "credit", "amount_cents": amt}]


def send(qa, method, path, **opts):
    if qa.source_error:
        return
    try:
        qa.api = books.request(method, path, **opts)
    except ApiUnreachable as err:
        qa.source_error = str(err)


def check(qa, description, fn):
    if qa.source_error:
        qa.unobservable(description, "the source could not be reached -- " + qa.source_error)
        return
    passed, detail = fn()
    qa.check(description, passed, detail)


# A minted token is a JSON string value "usr_<...>"; match the quoted value, not
# the bare prefix, to avoid false positives on field names.
def leaks_token(obj, admin_token):
    text = json.dumps(obj or {})
    return re.search(r'"usr_[A-Za-z0-9_-]{8,}"', text) is not None or ('"' + admin_token + '"') in text


# ---------------------------------------------------------------- adversarial requests


@when("an entry is posted with no token")
def entry_no_token(qa):
    send(qa, "POST", "/journal-entries", body={"entry_date": books.biz_day, "lines": _two_line(1, 9, 1000)})


@when("an entry is posted with a forged token")
def entry_forged_token(qa):
    send(qa, "POST", "/journal-entries", token=FORGED, body={"entry_date": books.biz_day, "lines": _two_line(1, 9, 1000)})


@when(parsers.parse("entry {entry_id:d} is read with no token"))
def entry_read_no_token(qa, entry_id):
    send(qa, "GET", "/journal-entries/" + str(entry_id))


@when("the admin entry list is read with the accountant's token")
def admin_list_accountant(qa):
    send(qa, "GET", "/admin/entries", token=books.seed_book)


@when("the admin overview is read with the accountant's token")
def admin_overview_accountant(qa):
    send(qa, "GET", "/admin/overview", token=books.seed_book)


@when("the admin overview is read with a token that extends the admin token")
def admin_overview_extended(qa):
    send(qa, "GET", "/admin/overview", token=books.admin_token + "x")


@when("the trial balance is read with no token")
def trial_balance_no_token(qa):
    send(qa, "GET", "/reports/trial-balance")


@when("the chart of accounts is read with no token")
def accounts_no_token(qa):
    send(qa, "GET", "/accounts")


@when("an accountant registers")
def accountant_registers(qa):
    send(qa, "POST", "/users", body={"name": "Sec Accountant"})


# ---------------------------------------------------------------- generic assertions


@then(parsers.parse('the response is an error with code "{code}"'))
def response_error_code(qa, code):
    check(qa, "error code " + code, lambda: (bool(qa.api) and qa.api["body"] and qa.api["body"].get("code") == code, json.dumps(qa.api["body"]) if qa.api and qa.api["body"] else "no body"))


@then("the response carries a token")
def response_carries_token(qa):
    def ev():
        t = (qa.api and qa.api["body"] or {}).get("token")
        return (isinstance(t, str) and len(t) > 0, "token " + ("present" if t else "absent"))
    check(qa, "response carries a token", ev)


# ---------------------------------------------------------------- secret hygiene


@then("the entry response carries no bearer token")
def entry_no_bearer(qa):
    check(qa, "entry body has no token", lambda: (not leaks_token(qa.entry, books.admin_token), "TOKEN LEAKED" if leaks_token(qa.entry, books.admin_token) else "clean"))


@then("the accounts response carries no bearer token")
def accounts_no_bearer(qa):
    check(qa, "accounts body has no token", lambda: (not leaks_token(qa.api and qa.api["body"], books.admin_token), "TOKEN LEAKED" if leaks_token(qa.api and qa.api["body"], books.admin_token) else "clean"))


@then("the report response carries no bearer token")
def report_no_bearer(qa):
    check(qa, "report body has no token", lambda: (not leaks_token(qa.api and qa.api["body"], books.admin_token), "TOKEN LEAKED" if leaks_token(qa.api and qa.api["body"], books.admin_token) else "clean"))
